Support `href`/`text` element filters

Twixes opened the pull request:
## Changes

Resolves #5598.

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests

Twixes committed:
Fix `icontains` and support `href/`text` element filters

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
+                else:
+                    attributes[key] = [re.escape(text) for text in vals]
+    is_negated = operator in NEGATED_OPERATORS
+    if attributes:
```
Relying on python empty objects being falsy is a bit shaky. Better to be explicit here.

Twixes replied:
It's not shaky, it's the "pythonic" way of doing this.

Twixes replied:
It's not shaky, it's the "pythonic" way of doing this.

macobo replied:
Hmm, checked out [pep8](https://www.python.org/dev/peps/pep-0008/). Fair regarding sequences: 

`For sequences, (strings, lists, tuples), use the fact that empty sequences are false:`

This does not mention dicts but w/e. :) Might be a footgun in there if using your own collections class.

But keep this also in mind then:

```
Comparisons to singletons like None should always be done with is or is not, never the equality operators.

Also, beware of writing if x when you really mean if x is not None -- e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!
```

We violate this just above in this file.

Twixes replied:
Yeah, the lacking null checks are a fair point, improved that.

Twixes committed:
Test `is_set`/`is_not_set` too

Twixes committed:
Require newer `flake8` to support Py 3.8 walrus operator

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
-
-    if len(attributes.keys()) > 0:
-        or_conditions = []
+        if vals := filters.get(key):
```
Code style nit: IMO relying in None here is shaky and makes this hard to follow since vals is only used N lines down.

How about:

```
if key in filters
```

However looking at the code a bit more I think it could benefit from extracting a function.

```
if key in filters:
  attributes[key] = get_element_match(filters[key], operator or "exact")

def get_element_match(values: Union[str, List[str]], operator: OperatorType):
  # Do the if-else here
```

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
+                    attributes[key] = vals
+                else:
+                    attributes[key] = [re.escape(text) for text in vals]
+    is_negated = operator in NEGATED_OPERATORS
```
Style nit: I'd inline this given it's used N lines down and breaks up the flow of reading this complicated function.

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
-                or_conditions.append("match(elements_chain, %({}_{}_{}_attributes_regex)s)".format(prepend, key, idx))
-            if len(or_conditions) > 0:
-                conditions.append("(" + (" OR ".join(or_conditions)) + ")")
+                optional_flag = "(?i)" if operator.endswith("icontains") else ""
```
This is out of sync with what we were doing in property.py.

We should use `ILIKE` here as well instead of a custom solution.

Twixes replied:
We already use regex for everything in here, I think this significantly simpler and less error-prone without an extra branch for `(not_)icontains`-only `ILIKE`.

macobo replied:
See the main comment I made around this - I think this is scope creep and shouldn't be in this PR.


That said, a proper solution would still unify this with `properties.py` in terms of approaches to icontains. Having similar, but different approaches is just a recipe for bugs/inconsistencies down the line.

Twixes replied:
Normal property filtering involves values from parsed JSON though, while in this case we use the elements directly, so the benefit of a regex is that i.e. we can handle double quote characters better.

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
+    if conditions:
+        final_conditions = f"{'NOT ' if is_negated else ''}({' OR '.join(conditions)})"
+    else:
+        final_conditions = "" if is_negated else "0 = 192"
```
This does not match the previous behavior - this is a new bug!

Twixes replied:
I don't think I see the bug, could you set out an example?

macobo replied:
The behavior changed when calling `filter_element` from `format_action_filter` on an autocapture event. Having no conditions is completely valid there - previously this function returned an empty string, now it returns `0 = 1`

Twixes replied:
Right – made things smarter.

macobo replied:
+1 good catch re behavior in models/property here :sweat_smile:

macobo reviewed (changes requested):
The resulting code is hard to follow. Pointed out few suggestions.

Meta-feedback: Your PRs tend to scope-creep. In this case you also started "fixing" a bunch of logic around whether the function should return an empty filter vs `0 = 1` and how icontains works. It's made worse since these are implicit changes - not mentioned in the linked ticket, not mentioned in the PR itself.

These kind of changes make PRs painful to review, hard to verify whether all the cases are covered and make it more likely for bugs slip through the cracks. For the author it also makes it harder to get the darn thing green and in.

I'd suggest doing things in separate PRs - one for the bug fix and one for the behavioral change.

Twixes committed:
Better handle empty mutlivalue operators and support doublequotes

Twixes committed:
Polish all element filtering cases

Twixes committed:
Roll back general `icontains` fix

Twixes commented:
I agree the `icontains` fix that applies more generally (but is literally a single letter added in two places) could be a separate PR. Disagree with this being scope creep besides that though – the problem here was that there was NO previous behavior – according to the UI these filters were perfectly possible, but in reality didn't do a thing. Hence I just added support for them to avoid any further confusion by users (like in the linked issue).
I added a bunch of tests here, so we have sensible coverage of the feature, and everything is green too.

macobo commented on ee/clickhouse/models/property.py:
```diff
@@ -225,55 +229,94 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
 
 
-def filter_element(filters: Dict, prepend: str = "") -> Tuple[str, Dict]:
+def process_ok_values(ok_values: Any, operator: OperatorType) -> List[str]:
```
Nittiest nit: I like placing functions like this at the end of a file - more important stuff at the top. This doesn't really matter though :sweat_smile:

Twixes replied:
I remember we talked about that at the offsite in October 😄 It's a fair point readability-wise

macobo reviewed (approved):
Good work! Thanks for fixing this :)

Twixes committed:
Make minor improvements
